File: graph-coloring.js
//Class Graph
class Graph 
{
    // Constructor
    constructor(v) 
    {
        this.V = v;

        this.adj = new Array(v);

        for(let i = 0; i < v; ++i) 

        {

            this.adj[i] = [];
            
        }
    }

    //add Edge method
    addEdge(v,w) 
    {
        this.adj[v].push(w);

         this.adj[w].push(v);
    }

    //Graph coloring method
    graphColoring() 
    {

        let result = new Array(this.V);
        
        let available = new Array(this.V);

        // Initialize all vertices as unassigned
        for(let i = 0; i < this.V; i++) 

        {

            result[i] = -1;

            available[i] = true;

        }

        //assign the first color
        result[0]  = 0;

        // Assign colors to remaining V-1 vertices
        for(let u = 1; u < this.V; u++) 

        {

            // Process all adjacent vertices and flag their colors as unavailable
            for(let it in this.adj[u]) 
        
            {
        
                let i = it;
        
                if (result[i] != -1) {
        
                    available[result[i]] = false;
        
                }
        
            }

            // Find the first available color
            for(let color = 0; color < this.V; color++) 

            {
                if (available[color]) 
                
                {

                    result[u] = color;

                    break;

                }
            }

            // Reset the values back to true
            for(let i = 0; i < this.V; i++) 
            {

                available[i] = true;

            }

        }

        // print the result
        for(let u = 0; u < this.V; u++) 

        {

            console.log(`Vertex ${u} ->  Color ${result[u]}`)

        }
    }
}

let graph = new Graph(5);
graph.addEdge(0, 1);
graph.addEdge(0, 2);
graph.addEdge(1, 2);
graph.addEdge(1, 3);
graph.addEdge(2, 3);
graph.addEdge(3, 4);
graph.graphColoring();

File: graph-coloring.cpp
// A C++ program to implement greedy algorithm for graph coloring
#include <iostream>
#include <list>
using namespace std;

//class Graph
class Graph 
{
    int V;
    list<int> *adj;
public:
    
    // Constructor and destructor
    Graph(int V) 
    {

        this->V = V;

        adj = new list<int>[V];
    }

    ~Graph() 
    { 
        delete[] adj; 
    }

    //add edge method
    void addEdge(int v, int w)
    {
        adj[v].push_back(w);

        adj[w].push_back(v); 
    }

    //graph coloring method
    void graphColoring()
    {
        int result[V];

        bool available[V];

        // Initialize all vertices as unassigned
        for (int u = 1; u < V; u++)
        {
            result[u] = -1; 

            available[u] = false;
        }

        // assign the first color
        result[0] = 0;

        // Assign colors to remaining V-1 vertices
        for (int u = 1; u < V; u++)
        {
            // Process all adjacent vertices and flag their colors as unavailable
            list<int>::iterator i;

            for (i = adj[u].begin(); i != adj[u].end(); ++i)
            {

                if (result[*i] != -1) 
                    {
                        available[result[*i]] = true;
                    }
            }

            // Find the first available color
            for (int color = 0; color < V; color++) 
            {
                if (available[color] == false) 
                {
                    result[u] = color; 
                    break;
                }
            }

            // Reset the values back to false for the next iteration
            for (i = adj[u].begin(); i != adj[u].end(); ++i) 
            {
                if (result[*i] != -1) 
                {
                    available[result[*i]] = false;
                }
            }
        }

        // print the result
        for (int u = 0; u < V; u++) 
        {
            cout << "Vertex " << u << " ->  Color " << result[u] << endl;
        }
    }
};

int main()
{
    Graph graph(5);
    graph.addEdge(0, 1);
    graph.addEdge(0, 2);
    graph.addEdge(1, 2);
    graph.addEdge(1, 3);
    graph.addEdge(2, 3);
    graph.addEdge(3, 4);
    graph.graphColoring();

    return 0;
}